kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/metrics/MetricsCollector.java

```diff
@@ -0,0 +1,387 @@
+    }
+
+    /**
+     * Gets component name.
+     *
+     * @return the component name
+     */
+    public String getComponentName() {
+        return componentName;
+    }
+
+    /**
+     * Gets metrics path.
+     *
+     * @return the metrics path
+     */
+    public String getMetricsPath() {
+        return metricsPath;
+    }
+
+    /**
+     * Gets metrics port.
+     *
+     * @return the metrics port
+     */
+    public int getMetricsPort() {
+        return metricsPort;
+    }
+
+    /**
+     * Gets collected data.
+     *
+     * @return the collected data
+     */
+    public Map<String, String> getCollectedData() {
+        return collectedData;
+    }
+
+    /**
+     * New builder metrics collector . builder.
+     *
+     * @return the metrics collector . builder
+     */
+    protected MetricsCollector.Builder newBuilder() {
+        return new MetricsCollector.Builder();
+    }
+
+    /**
+     * Update builder metrics collector . builder.
+     *
+     * @param builder the builder
+     * @return the metrics collector . builder
+     */
+    protected MetricsCollector.Builder updateBuilder(MetricsCollector.Builder builder) {
+        return builder
+                .withNamespaceName(getNamespaceName())
+                .withComponentName(getComponentName())
+                .withComponentType(getComponentType())
+                .withScraperPodName(getScraperPodName());
+    }
+
+    /**
+     * To builder metrics collector . builder.
+     *
+     * @return the metrics collector . builder
+     */
+    public MetricsCollector.Builder toBuilder() {
+        return updateBuilder(newBuilder());
+    }
+
+    /**
+     * Instantiates a new Metrics collector.
+     *
+     * @param builder the builder
+     */
+    protected MetricsCollector(Builder builder) {
+        if (builder.namespaceName == null || builder.namespaceName.isEmpty())
+            builder.namespaceName = kubeClient().getNamespace();
+        if (builder.scraperPodName == null || builder.scraperPodName.isEmpty())
+            throw new InvalidParameterException("Scraper Pod name is not set");
+        if (builder.componentType == null)
+            throw new InvalidParameterException("Component type is not set");
+
+        componentType = builder.componentType;
+
+        if (builder.metricsPort <= 0)
+            builder.metricsPort = getDefaultMetricsPort();
+        if (builder.metricsPath == null || builder.metricsPath.isEmpty())
+            builder.metricsPath = getDefaultMetricsPath();
+
+        namespaceName = builder.namespaceName;
+        scraperPodName = builder.scraperPodName;
+        metricsPort = builder.metricsPort;
+        metricsPath = builder.metricsPath;
+        componentName = builder.componentName;
+        componentLabelSelector = getLabelSelectorForResource();
+    }
+
+    private LabelSelector getLabelSelectorForResource() {
+        if (Objects.requireNonNull(this.componentType) == ComponentType.Kroxylicious) {
```

Comment:
ouch! that violates least surprise. by about a parsec. 

etters shouldn't make assertions about state, even if they are private.  Best to impose the condition on construction, so it is easy for the caller to know where they went wrong.